File: Adhiti_Shaili/Day 16/Majority Element.cpp
#include<bits/stdc++.h>
using namespace std;

int majorityElement(int a[], int size)
{
    int count=1,val=a[0];
    for(int i=1;i<size;i++)
    {
        if(a[i]==val)
        {
            count++;
        }
        else if(count==1)
        {
            val=a[i];
            count=1;
        }
        else
        {
            count--;
        }
    }
    count=0;
    for(int i=0;i<size;i++)
    {
        if(a[i]==val)
        {
            count++;
        }
    }
    if(count>size/2)
    {
        return val;
    }
    return -1;
    
}

int main(){

    int t;
    cin >> t;

    while(t--){
        int n;
        cin >> n;
        int arr[n];
        
        for(int i = 0;i<n;i++){
            cin >> arr[i];
        }
        
        cout << majorityElement(arr, n) << endl;
    }

    return 0;
}

File: sanket-dalvi/Day16/majorityElement.py
def majorityElement(A, N):
    if not N:
        return -1
    i = 1
    majElem = A[0]
    majCount = 1
    while i < N:
        if majCount == 0:
            majCount = 1
            majElem = A[i]
        if A[i] == majElem:
            majCount += 1
        else:
            majCount -= 1
        i += 1
    if majCount <= 1:
        return -1
    else:
        count = 0
        for i in A:
            if i == majElem:
                count += 1
    return count if count >= N//2 else -1


# Test
arr = [3, 1, 7, 8, 1, 3, 3, 2]
print(majorityElement(arr, len(arr)))
